File: tools/aegis-rig/deform.py
#!/usr/bin/env python3
"""Deformation probe — Packet Beta Module 3, over a real rigged asset.

AUTHORIZE proves a rig is *valid*. This asks whether it is any *good*, by
posing it and looking for the artifacts a rigger actually looks for:
pinching, inverted faces, tearing, and parts driven into each other.

The last one is the interesting one on this corpus. A gasket rigged at 8
joints is 374 rigid shells hanging off a chain; bend the chain and those
shells swing into each other long before any single shell deforms badly.
That is the failure this corpus is actually prone to, and it is invisible
to a weight-sum check.

Python builds the pose matrices — a handful of 4x4s, where the language
does not matter — and the Rust kernel does the per-vertex work, per
doctrine 015.
"""

from __future__ import annotations

import json
import math
import struct
import subprocess
import sys
from array import array
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "aegis-inspect"))
import inspector  # noqa: E402
import solver  # noqa: E402

DEFORM_MAGIC = b"ADEF"
DEFORM_VERSION = 1

# Four bends, shallow to severe. A rig that survives 15 degrees and fails at
# 60 is telling you where its usable range ends, which is more useful than a
# single pass/fail at one arbitrary angle.
DEFAULT_ANGLES = (15.0, 30.0, 45.0, 60.0)


class DeformError(RuntimeError):
    pass


# --- small matrix helpers (column-major, glTF convention) -------------------

IDENTITY = (1.0, 0, 0, 0, 0, 1.0, 0, 0, 0, 0, 1.0, 0, 0, 0, 0, 1.0)


def mat_mul(a, b):
    """a @ b, both column-major 4x4 as flat 16-tuples."""
    out = [0.0] * 16
    for col in range(4):
        for row in range(4):
            out[col * 4 + row] = sum(a[k * 4 + row] * b[col * 4 + k] for k in range(4))
    return tuple(out)


def mat_translate(t):
    return (1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, t[0], t[1], t[2], 1)


def mat_rotate(axis: int, angle: float):
    """Rotation about a world axis, by index."""
    c, s = math.cos(angle), math.sin(angle)
    if axis == 0:
        return (1, 0, 0, 0, 0, c, s, 0, 0, -s, c, 0, 0, 0, 0, 1)
    if axis == 1:
        return (c, 0, -s, 0, 0, 1, 0, 0, s, 0, c, 0, 0, 0, 0, 1)
    return (c, s, 0, 0, -s, c, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1)


def node_local(node: dict):
    if "matrix" in node:
        return tuple(node["matrix"])
    local = mat_translate(node.get("translation", (0.0, 0.0, 0.0)))
    scale = node.get("scale")
    if scale:
        local = mat_mul(local, (scale[0], 0, 0, 0, 0, scale[1], 0, 0,
                                0, 0, scale[2], 0, 0, 0, 0, 1))
    return local


# --- reading a rigged asset -------------------------------------------------

def _read_vec4(gltf, blob, accessor_index, kind):
    accessor = gltf["accessors"][accessor_index]
    view = gltf["bufferViews"][accessor["bufferView"]]
    start = view.get("byteOffset", 0) + accessor.get("byteOffset", 0)
    count = accessor["count"]
    component = accessor["componentType"]
    if kind == "joints":
        fmt = {5121: "B", 5123: "H"}.get(component)
        if fmt is None:
            raise DeformError(f"unsupported JOINTS_0 component type {component}")
        raw = struct.unpack_from(f"<{count * 4}{fmt}", blob, start)
        return array("H", raw)
    if component != 5126:
        raise DeformError("WEIGHTS_0 must be float for this probe")
    return array("f", struct.unpack_from(f"<{count * 4}f", blob, start))


def load_rig(path: Path) -> dict:
    gltf, blob, _ = solver._split_glb(path.read_bytes())
    skins = gltf.get("skins") or []
    if not skins:
        raise DeformError("asset has no skin — nothing to pose")
    skin = skins[0]

    prim = None
    for mesh in gltf.get("meshes", []):
        for candidate in mesh.get("primitives", []):
            if "JOINTS_0" in candidate.get("attributes", {}):
                prim = candidate
                break
        if prim:
            break
    if prim is None:
        raise DeformError("no primitive carries JOINTS_0")

    positions = solver._read_positions(gltf, blob, prim["attributes"]["POSITION"])
    indices = inspector._read_indices(gltf, blob, 0, prim["indices"])
    joint_ids = _read_vec4(gltf, blob, prim["attributes"]["JOINTS_0"], "joints")
    weights = _read_vec4(gltf, blob, prim["attributes"]["WEIGHTS_0"], "weights")

    nodes = gltf.get("nodes", [])
    joint_nodes = skin["joints"]

    parent_of = {}
    for index, node in enumerate(nodes):
        for child in node.get("children", []):
            parent_of[child] = index

    def world(node_index):
        chain = []
        current = node_index
        while current is not None:
            chain.append(current)
            current = parent_of.get(current)
        matrix = IDENTITY
        for index in reversed(chain):
            matrix = mat_mul(matrix, node_local(nodes[index]))
        return matrix

    rest_world = [world(n) for n in joint_nodes]

    ibm_index = skin.get("inverseBindMatrices")
    if ibm_index is None:
        inverse_bind = [IDENTITY] * len(joint_nodes)
    else:
        accessor = gltf["accessors"][ibm_index]
        view = gltf["bufferViews"][accessor["bufferView"]]
        start = view.get("byteOffset", 0) + accessor.get("byteOffset", 0)
        flat = struct.unpack_from(f"<{accessor['count'] * 16}f", blob, start)
        inverse_bind = [tuple(flat[i * 16:(i + 1) * 16]) for i in range(accessor["count"])]

    return {
        "positions": positions,
        "indices": indices,
        "joint_ids": joint_ids,
        "weights": weights,
        "joint_nodes": joint_nodes,
        "rest_world": rest_world,
        "inverse_bind": inverse_bind,
        "parent_of": parent_of,
        "nodes": nodes,
    }


# --- posing -----------------------------------------------------------------

def build_poses(rig: dict, angles=DEFAULT_ANGLES) -> tuple[list, int]:
    """Bend the chain progressively: the root holds still and each joint down
    the chain adds another share of the angle, which is what a limb does."""
    joint_nodes = rig["joint_nodes"]
    origins = [(m[12], m[13], m[14]) for m in rig["rest_world"]]

    # Chain direction from the joint positions; rotate about a world axis
    # perpendicular to it, so the bend is visible rather than a twist.
    spread = [max(p[i] for p in origins) - min(p[i] for p in origins) for i in range(3)]
    chain_axis = spread.index(max(spread))
    bend_axis = (chain_axis + 2) % 3

    parent_of = rig["parent_of"]
    index_of = {node: i for i, node in enumerate(joint_nodes)}
    count = len(joint_nodes)

    poses = []
    for degrees in angles:
        total = math.radians(degrees)
        posed_world = [None] * count
        for i, node in enumerate(joint_nodes):
            share = total / max(count - 1, 1)
            local = node_local(rig["nodes"][node])
            # Rotate in the joint's own frame: translate out, spin, translate
            # back is implicit because the local matrix already carries the
            # offset from the parent.
            local = mat_mul(local, mat_rotate(bend_axis, share if i > 0 else 0.0))
            parent_node = parent_of.get(node)
            parent_index = index_of.get(parent_node)
            if parent_index is None:
                posed_world[i] = local
            else:
                posed_world[i] = mat_mul(posed_world[parent_index], local)
        poses.append([mat_mul(posed_world[i], rig["inverse_bind"][i])
                      for i in range(count)])
    return poses, bend_axis


# --- the probe --------------------------------------------------------------

def probe(path: Path, angles=DEFAULT_ANGLES) -> dict:
    if not solver.core_available():
        raise DeformError("rigging core is not built — run cargo build --release")

    rig = load_rig(path)
    poses, bend_axis = build_poses(rig, angles)

    vertices = len(rig["positions"])
    joints = len(rig["joint_nodes"])
    header = DEFORM_MAGIC + struct.pack(
        "<IIIII", DEFORM_VERSION, vertices, len(rig["indices"]), joints, len(poses))

    pos = array("f")
    for p in rig["positions"]:
        pos.extend(p)
    idx = array("I", rig["indices"])
    mats = array("f")
    for pose in poses:
        for matrix in pose:
            mats.extend(matrix)

    payload = (header + pos.tobytes() + idx.tobytes()
               + rig["joint_ids"].tobytes() + rig["weights"].tobytes()
               + mats.tobytes())

    result = subprocess.run([str(solver.CORE_BINARY)], input=payload,
                            stdout=subprocess.PIPE, check=False)
    out = result.stdout
    if len(out) < 4:
        raise DeformError("deform kernel returned nothing")
    json_len = struct.unpack_from("<I", out, 0)[0]
    report = json.loads(out[4:4 + json_len].decode("utf-8"))
    if not report.get("ok"):
        raise DeformError(report.get("error", "deform kernel failed"))

    triangles = report["triangles"]
    graded = []
    for degrees, pose in zip(angles, report["poses"]):
        graded.append({**pose, "degrees": degrees, **grade(pose, triangles)})

    worst = max(graded, key=lambda g: ("fail", "warn", "pass").index(g["verdict"]) * -1)
    return {
        "ok": True,
        "asset": path.name,
        "triangles": triangles,
        "shells": report["shells"],
        "joints": joints,
        "bend_axis": "XYZ"[bend_axis],
        "poses": graded,
        "verdict": worst["verdict"],
        "note": worst["reason"],
    }


def grade(pose: dict, triangles: int) -> dict:
    """Thresholds as fractions of the mesh, not absolute counts — otherwise a
    million-triangle asset fails for artifacts a small one is forgiven."""
    if triangles <= 0:
        return {"verdict": "fail", "reason": "no triangles to measure"}

    inverted = pose["inverted"] / triangles
    collapsed = pose["collapsed"] / triangles
    torn = pose["torn"] / max(triangles * 3, 1)

    if inverted > 0.001:
        return {"verdict": "fail",
                "reason": f"{pose['inverted']:,} inverted faces "
                          f"({inverted * 100:.2f}%) — surface turns inside out"}
    if collapsed > 0.05:
        return {"verdict": "fail",
                "reason": f"{pose['collapsed']:,} collapsed faces "
                          f"({collapsed * 100:.1f}%) — severe pinching"}
    if pose["clipping_pairs"] > 0:
        return {"verdict": "warn",
                "reason": f"{pose['clipping_pairs']:,} shell pairs newly "
                          "overlapping — parts driven into each other"}
    if collapsed > 0.01 or torn > 0.01:
        return {"verdict": "warn",
                "reason": f"{pose['collapsed']:,} pinched, {pose['torn']:,} torn edges"}
    return {"verdict": "pass",
            "reason": f"no inversion, no clipping, max stretch "
                      f"{pose['max_stretch']:.2f}x"}


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print("usage: deform.py <rigged.glb> [angle ...]")
        return 2
    angles = tuple(float(a) for a in argv[2:]) or DEFAULT_ANGLES
    print(json.dumps(probe(Path(argv[1]), angles), indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: tools/aegis-rig/rust/src/deform.rs
//! Deformation probe — Packet Beta Module 3, the clipping-and-pinching test.
//!
//! A rig that passes AUTHORIZE is *valid*: weights sum to 1.0, no shell is
//! split across bones. That says nothing about whether it deforms *well*.
//! This module poses the rig and measures what breaks.
//!
//! Four failure modes, each a real artifact a rigger looks for:
//!
//! - **Pinching** — a triangle collapses toward zero area as bones pull its
//!   vertices together. The classic candy-wrapper symptom.
//! - **Inversion** — a triangle's normal flips. The surface has turned inside
//!   out; worse than pinching because it renders as a hole.
//! - **Stretch** — an edge grows far beyond its rest length. Tearing.
//! - **Clipping** — two shells that were disjoint in the rest pose now
//!   overlap. On this corpus that is *the* risk: 374 rigid shells rotating
//!   about a chain will drive into each other long before any single shell
//!   deforms badly.
//!
//! Python builds the pose matrices — a handful of 4x4s, where performance is
//! irrelevant — and this does the per-vertex and per-triangle work.

/// A triangle below this fraction of its rest area counts as pinched.
const COLLAPSE_RATIO: f64 = 0.25;
/// An edge above this multiple of its rest length counts as torn.
const STRETCH_LIMIT: f64 = 2.0;
/// Overlap beyond this fraction of the smaller part's volume is real
/// interpenetration rather than two parts legitimately closing on each other.
const DEEP_OVERLAP: f64 = 0.5;

pub struct DeformInput {
    pub positions: Vec<[f64; 3]>,
    pub indices: Vec<u32>,
    pub joint_ids: Vec<u16>,
    pub weights: Vec<f32>,
    /// Per pose, per joint: the 4x4 skinning matrix, column-major.
    pub poses: Vec<Vec<[f64; 16]>>,
    pub labels: Vec<u32>,
    pub shell_count: usize,
}

pub struct PoseReport {
    pub collapsed: usize,
    pub inverted: usize,
    pub torn: usize,
    pub max_stretch: f64,
    pub mean_stretch: f64,
    pub max_displacement: f64,
    /// Pairs whose boxes newly touch — includes legitimate articulation.
    pub clipping_pairs: usize,
    /// Pairs genuinely interpenetrating. This is the defect count.
    pub deep_clipping: usize,
}

fn apply(m: &[f64; 16], p: &[f64; 3]) -> [f64; 3] {
    [
        m[0] * p[0] + m[4] * p[1] + m[8] * p[2] + m[12],
        m[1] * p[0] + m[5] * p[1] + m[9] * p[2] + m[13],
        m[2] * p[0] + m[6] * p[1] + m[10] * p[2] + m[14],
    ]
}

fn cross(a: [f64; 3], b: [f64; 3]) -> [f64; 3] {
    [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]]
}

fn sub(a: [f64; 3], b: [f64; 3]) -> [f64; 3] {
    [a[0] - b[0], a[1] - b[1], a[2] - b[2]]
}

fn dot(a: [f64; 3], b: [f64; 3]) -> f64 {
    a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
}

fn norm(a: [f64; 3]) -> f64 {
    dot(a, a).sqrt()
}

/// Linear blend skinning. Packet Beta names LBS vs dual-quaternion as the
/// real axis of choice; with at most two influences and no twist in the rest
/// pose, LBS is what the solver's output actually implies, so it is what the
/// probe measures. When a twisting joint appears, this is the function that
/// has to grow a second implementation.
fn skin(input: &DeformInput, pose: &[[f64; 16]]) -> Vec<[f64; 3]> {
    let mut out = Vec::with_capacity(input.positions.len());
    for (v, p) in input.positions.iter().enumerate() {
        let mut acc = [0f64; 3];
        for k in 0..4 {
            let w = input.weights[4 * v + k] as f64;
            if w == 0.0 {
                continue;
            }
            let j = input.joint_ids[4 * v + k] as usize;
            if j >= pose.len() {
                continue;
            }
            let moved = apply(&pose[j], p);
            acc[0] += w * moved[0];
            acc[1] += w * moved[1];
            acc[2] += w * moved[2];
        }
        out.push(acc);
    }
    out
}

fn shell_boxes(positions: &[[f64; 3]], labels: &[u32], shell_count: usize)
    -> Vec<([f64; 3], [f64; 3])> {
    let mut boxes = vec![([f64::INFINITY; 3], [f64::NEG_INFINITY; 3]); shell_count];
    for (v, p) in positions.iter().enumerate() {
        let s = labels[v] as usize;
        for i in 0..3 {
            boxes[s].0[i] = boxes[s].0[i].min(p[i]);
            boxes[s].1[i] = boxes[s].1[i].max(p[i]);
        }
    }
    boxes
}

fn overlaps(a: &([f64; 3], [f64; 3]), b: &([f64; 3], [f64; 3])) -> bool {
    (0..3).all(|i| a.0[i] <= b.1[i] && b.0[i] <= a.1[i])
}

fn overlap_volume(a: &([f64; 3], [f64; 3]), b: &([f64; 3], [f64; 3])) -> f64 {
    let mut v = 1.0;
    for i in 0..3 {
        let d = a.1[i].min(b.1[i]) - a.0[i].max(b.0[i]);
        if d <= 0.0 {
            return 0.0;
        }
        v *= d;
    }
    v
}

fn box_volume(a: &([f64; 3], [f64; 3])) -> f64 {
    (0..3).map(|i| (a.1[i] - a.0[i]).max(0.0)).product()
}

/// Shells that were apart in the rest pose and interpenetrate now, **graded
/// by how deeply**.
///
/// Counting any new box overlap was the first version of this and it was
/// wrong in a way worth recording: on a bending articulated model, adjacent
/// parts legitimately approach each other as a joint closes. Measured on
/// gasket at 45 degrees, 341 pairs "collided" but the median overlap was 8%
/// of the smaller part's volume — that is articulation, not clipping. Only
/// 29 pairs exceeded 50%.
///
/// So the number that means something is the deep count. `touching` is kept
/// because it is a useful denominator, not because it is a defect count.
fn clipping(rest: &[[f64; 3]], posed: &[[f64; 3]], labels: &[u32], shell_count: usize)
    -> (usize, usize) {
    if shell_count < 2 {
        return (0, 0);
    }
    let rest_boxes = shell_boxes(rest, labels, shell_count);
    let posed_boxes = shell_boxes(posed, labels, shell_count);
    let (mut touching, mut deep) = (0, 0);
    for a in 0..shell_count {
        for b in (a + 1)..shell_count {
            if !overlaps(&posed_boxes[a], &posed_boxes[b])
                || overlaps(&rest_boxes[a], &rest_boxes[b])
            {
                continue;
            }
            touching += 1;
            let smaller = box_volume(&posed_boxes[a]).min(box_volume(&posed_boxes[b]));
            if smaller > 0.0
                && overlap_volume(&posed_boxes[a], &posed_boxes[b]) / smaller > DEEP_OVERLAP
            {
                deep += 1;
            }
        }
    }
    (touching, deep)
}

pub fn probe(input: &DeformInput) -> Vec<PoseReport> {
    let rest = &input.positions;
    let mut reports = Vec::new();

    for pose in &input.poses {
        let posed = skin(input, pose);

        let (mut collapsed, mut inverted, mut torn) = (0usize, 0usize, 0usize);
        let (mut max_stretch, mut stretch_sum) = (0f64, 0f64);
        let mut edges = 0usize;

        for tri in input.indices.chunks_exact(3) {
            let (a, b, c) = (tri[0] as usize, tri[1] as usize, tri[2] as usize);
            if a >= rest.len() || b >= rest.len() || c >= rest.len() {
                continue;
            }
            let nr = cross(sub(rest[b], rest[a]), sub(rest[c], rest[a]));
            let np = cross(sub(posed[b], posed[a]), sub(posed[c], posed[a]));
            let (ar, ap) = (norm(nr), norm(np));

            if ar > 0.0 {
                if ap / ar < COLLAPSE_RATIO {
                    collapsed += 1;
                }
                if dot(nr, np) < 0.0 {
                    inverted += 1;
                }
            }

            for (x, y) in [(a, b), (b, c), (c, a)] {
                let lr = norm(sub(rest[y], rest[x]));
                if lr <= 0.0 {
                    continue;
                }
                let ratio = norm(sub(posed[y], posed[x])) / lr;
                if ratio > STRETCH_LIMIT {
                    torn += 1;
                }
                max_stretch = max_stretch.max(ratio);
                stretch_sum += ratio;
                edges += 1;
            }
        }

        let mut max_displacement = 0f64;
        for v in 0..rest.len() {
            max_displacement = max_displacement.max(norm(sub(posed[v], rest[v])));
        }

        let clip = clipping(rest, &posed, &input.labels, input.shell_count);
        reports.push(PoseReport {
            collapsed,
            inverted,
            torn,
            max_stretch,
            mean_stretch: if edges > 0 { stretch_sum / edges as f64 } else { 0.0 },
            max_displacement,
            clipping_pairs: clip.0,
            deep_clipping: clip.1,
        });
    }

    reports
}
